dkg/frost/frost.go

```diff
@@ -0,0 +1,756 @@
+package frost
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+
+	"github.com/bloxapp/ssv-spec/dkg"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/coinbase/kryptology/pkg/dkg/frost"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	ecies "github.com/ecies/go/v2"
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/pkg/errors"
+)
+
+var thisCurve = curves.BLS12381G1()
+
+func init() {
+	_ = bls.Init(bls.BLS12_381)
+	_ = bls.SetETHmode(bls.EthModeDraft07)
+}
+
+type FROST struct {
+	identifier   dkg.RequestID
+	network      dkg.Network
+	signer       types.DKGSigner
+	storage      dkg.Storage
+	threshold    uint32
+	currentRound DKGRound
+
+	operatorID types.OperatorID
+	operators  []uint32
+	party      *frost.DkgParticipant
+	sessionSK  *ecies.PrivateKey
+
+	msgs           map[DKGRound]map[uint32]*dkg.SignedMessage
+	operatorShares map[uint32]*bls.SecretKey
+}
+
+type DKGRound int
+
+const (
+	Preparation DKGRound = iota + 1
+	Round1
+	Round2
+	Blame
+)
+
+func New(
+	network dkg.Network,
+	operatorID types.OperatorID,
+	requestID dkg.RequestID,
+	signer types.DKGSigner,
+	storage dkg.Storage,
+) dkg.KeyGenProtocol {
+
+	msgs := make(map[DKGRound]map[uint32]*dkg.SignedMessage)
+	msgs[Preparation] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Round1] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Round2] = make(map[uint32]*dkg.SignedMessage)
+	msgs[Blame] = make(map[uint32]*dkg.SignedMessage)
+
+	return &FROST{
+		identifier: requestID,
+		network:    network,
+		signer:     signer,
+		storage:    storage,
+		operatorID: operatorID,
+
+		msgs:           msgs,
+		operatorShares: make(map[uint32]*bls.SecretKey),
+	}
+}
+
+func (fr *FROST) Start(init *dkg.Init) error {
+	otherOperators := make([]uint32, 0)
+	for _, operatorID := range init.OperatorIDs {
+		if fr.operatorID == operatorID {
+			continue
+		}
+		otherOperators = append(otherOperators, uint32(operatorID))
+	}
+
+	operators := []uint32{uint32(fr.operatorID)}
+	operators = append(operators, otherOperators...)
+	fr.operators = operators
+
+	pctx := make([]byte, 16)
+	_, err := rand.Read(pctx)
+	if err != nil {
+		return err
+	}
+
+	party, err := frost.NewDkgParticipant(uint32(fr.operatorID), uint32(len(operators)), string(pctx), thisCurve, otherOperators...)
+	if err != nil {
+		return err
+	}
+
+	fr.party = party
+	fr.threshold = uint32(init.Threshold)
+
+	k, err := ecies.GenerateKey()
+	if err != nil {
+		return err
+	}
+	fr.sessionSK = k
+
+	fr.currentRound = Preparation
+
+	protocolMessage := &ProtocolMsg{
+		Round: Preparation,
+		PreparationMessage: &PreparationMessage{
+			SessionPk: k.PublicKey.Bytes(true),
+		},
+	}
+
+	protocolMessageBytes, err := protocolMessage.Encode()
```

Comment:
L119-L150 is probably used elsewhere so why not make a reusable func for broadcasting?